File: community.variability/python/community_variability/leave_one_out.py
"""Sensitivity analyses for community arrays."""

from __future__ import annotations

from collections.abc import Callable, Sequence

import numpy as np
import pandas as pd

from .community_metrics import CommunityArray, calc_metacommunity_metrics


def leave_one_out(
    X: CommunityArray | np.ndarray,
    margin: str | int,
    metric_fn: Callable[[CommunityArray | np.ndarray], pd.DataFrame] = calc_metacommunity_metrics,
) -> pd.DataFrame:
    """Recalculate metrics after removing each timestep, site, or taxon slice."""
    values = X.values if isinstance(X, CommunityArray) else X
    values = np.asarray(values, dtype=float)
    if values.ndim != 3:
        raise ValueError("Community data must have shape X[time, site, taxon].")

    if isinstance(margin, str):
        names = ["timestep", "site", "taxon"]
        if margin not in names:
            raise ValueError(f"margin must be one of {names}, not {margin!r}.")
        margin_idx = names.index(margin)
    else:
        margin_idx = int(margin)
        if margin_idx not in (0, 1, 2):
            raise ValueError("Integer margin must be 0, 1, or 2.")

    margin_name = ["timestep", "site", "taxon"][margin_idx]
    removed_col = f"{margin_name}_removed"
    labels = X.dimnames[margin_name] if isinstance(X, CommunityArray) else [str(i) for i in range(values.shape[margin_idx])]

    baseline = metric_fn(X)
    baseline.insert(0, removed_col, "Baseline")
    rows = [baseline]

    for label_idx, label in enumerate(labels):
        # Omit one ecological component, then recompute all metrics.
        # delta = metric_without_component - metric_baseline is added later.
        keep_idx = np.array([i for i in range(len(labels)) if i != label_idx], dtype=int)
        indices: list[object] = [slice(None), slice(None), slice(None)]
        indices[margin_idx] = keep_idx
        omitted_values = values[tuple(indices)]

        if isinstance(X, CommunityArray):
            dimnames = X.dimnames
            sliced_labels: dict[str, list[str]] = {}
            for axis, name in enumerate(["timestep", "site", "taxon"]):
                if axis == margin_idx:
                    sliced_labels[name] = [dimnames[name][int(i)] for i in keep_idx]
                else:
                    sliced_labels[name] = list(dimnames[name])
            omitted_X: CommunityArray | np.ndarray = CommunityArray(
                values=omitted_values,
                timestep=sliced_labels["timestep"],
                site=sliced_labels["site"],
                taxon=sliced_labels["taxon"],
            )
        else:
            omitted_X = omitted_values

        omitted_metrics = metric_fn(omitted_X)
        omitted_metrics.insert(0, removed_col, label)
        rows.append(omitted_metrics)

    return pd.concat(rows, ignore_index=True)


def add_baseline_delta(
    sensitivity_results: pd.DataFrame,
    removed_col: str,
    baseline_label: str = "Baseline",
    group_cols: str | Sequence[str] | None = None,
) -> pd.DataFrame:
    """Add delta = estimate_without_component - estimate_baseline."""
    group_cols = [] if group_cols is None else ([group_cols] if isinstance(group_cols, str) else list(group_cols))
    join_cols = group_cols + ["varname"]
    baseline = (
        sensitivity_results.loc[sensitivity_results[removed_col] == baseline_label, join_cols + ["estimate"]]
        .rename(columns={"estimate": "baseline"})
        .copy()
    )
    out = sensitivity_results.merge(baseline, on=join_cols, how="left")
    out["delta"] = out["estimate"] - out["baseline"]
    out["abs_delta"] = out["delta"].abs()
    return out

File: community.variability/matlab/community_variability/leave_one_out.m
function out = leave_one_out(X, margin)
%LEAVE_ONE_OUT Recalculate metrics after removing one slice at a time.
%
% margin can be "timestep", "site", "taxon", or integer 1, 2, 3.
% This estimates sensitivity: delta = metric_without_slice - baseline.

if isstruct(X)
    values = double(X.values);
    labels = X;
else
    values = double(X);
    labels = struct();
    labels.timestep = string(1:size(values, 1));
    labels.site = string(1:size(values, 2));
    labels.taxon = string(1:size(values, 3));
end
if ndims(values) ~= 3
    error("Community data must have shape X(time, site, taxon).");
end

names = ["timestep", "site", "taxon"];
if isstring(margin) || ischar(margin)
    margin_idx = find(names == string(margin), 1);
    if isempty(margin_idx)
        error("margin must be timestep, site, taxon, 1, 2, or 3.");
    end
else
    margin_idx = double(margin);
    if ~ismember(margin_idx, [1, 2, 3])
        error("Integer margin must be 1, 2, or 3.");
    end
end

margin_name = names(margin_idx);
removed_col = margin_name + "_removed";
slice_labels = labels.(margin_name);

baseline = calc_metacommunity_metrics(struct("values", values, ...
    "timestep", labels.timestep, "site", labels.site, "taxon", labels.taxon));
baseline.(removed_col) = repmat("Baseline", height(baseline), 1);
baseline = movevars(baseline, removed_col, "Before", 1);
out = baseline;

for label_idx = 1:numel(slice_labels)
    keep_idx = setdiff(1:numel(slice_labels), label_idx, "stable");
    if margin_idx == 1
        omitted = struct("values", values(keep_idx, :, :), "timestep", labels.timestep(keep_idx), ...
            "site", labels.site, "taxon", labels.taxon);
    elseif margin_idx == 2
        omitted = struct("values", values(:, keep_idx, :), "timestep", labels.timestep, ...
            "site", labels.site(keep_idx), "taxon", labels.taxon);
    else
        omitted = struct("values", values(:, :, keep_idx), "timestep", labels.timestep, ...
            "site", labels.site, "taxon", labels.taxon(keep_idx));
    end

    metrics = calc_metacommunity_metrics(omitted);
    metrics.(removed_col) = repmat(string(slice_labels(label_idx)), height(metrics), 1);
    metrics = movevars(metrics, removed_col, "Before", 1);
    out = [out; metrics]; %#ok<AGROW>
end
end
